### PR TITLE
huggingface-cli upload - Validate README.md before file hashing

### LINKED ISSUE #2451: `huggingface-cli upload` - Validate `README.md` before file hashing
For large datasets file hashing can take some time.

[Dataset Card validation](https://github.com/huggingface/huggingface_hub/blob/dfd73c05c7777804b14e9e3cec17e38b23c52ebe/src/huggingface_hub/hf_api.py#L3795-L3819) happens after file hashing.

"It's better to fail early than to fail after all the files have been ~~uploaded~~ hashed."

### PR DESCRIPTION
**Pull request by hlky**
- `README.md` validation is moved from `create_commit` to `_prepare_upload_folder_additions`.
- `_prepare_upload_folder_additions` is moved into `HfApi` class, this provides access to `_build_hf_headers`.
- `repo_type` and `token` are added to `_prepare_upload_folder_additions`'s parameters, these are needed for the request to `/api/validate-yaml` endpoint.
- Result from `filter_repo_objects` stored in `filtered_repo_objects` (converted to a list, as it's a generator) to avoid recalculation as it is now used to find `README.md` then later to return the `CommitOperationAdd`s.
- If found, a temporary `CommitOperationAdd` is created for `README.md` to provide `as_file`.
- `break` added after `README.md` validation to avoid iterating other files.

Closes #2451

### CODE REVIEW

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:9229]: Could you move this logic to a separate method like this?

```py
def _validate_yaml(self, content: str, *, repo_type: Optional[str] = None, token: Union[bool, str, None] = None):
    # takes as input the raw content to validate
    # make the request + print warnings + raise on error
    (...)
```

This would make the code more readable by been self-contained and separated from the `_prepare_upload_folder_additions` logic.

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:9201]: And then, instead of the `for ... in ...: if ... == "README.md": .... ... ... break` logic, could you use `relpath_to_abspath` instead like this:

```py
if "README.md" in filtered_repo_objects:
    self._validate_yaml(
        content=relpath_to_abspath["README.md"].read_text(),
        repo_type=repo_type,
        token=token,
    )
```

This way, it's not even needed to build the `CommitOperationAdd` just to read the file.

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:3804]: This part has to be kept in `create_commit` as well with something like that:

```py
for addition in additions:
    if addition.path_in_repo == "README.md":
        with addition.as_file() as file:
            content = file.read().decode()
        self._validate_yaml(content, ...)
```

Not all commits are created with `upload_folder` so we cannot assume the check has already been done.

**Review by Wauplin — COMMENTED** [review]
Hi @hlky, thanks for the PR and detailed description. I've left a couple comments to address before merging. The two main things are:
- I'd prefer to separate the "yaml validation" from the "list and filter files" logic
- we have to validate the yaml in `create_commit` in any case (not a problem to do it twice if `upload_folder` has been called)

File: src/huggingface_hub/hf_api.py
Comment (hlky) [reply src/huggingface_hub/hf_api.py]: Thanks, much better to use `read_text` from pathlib and avoid the for loop.

File: src/huggingface_hub/hf_api.py
Comment (hlky) [reply src/huggingface_hub/hf_api.py]: Added back in, thanks.

File: src/huggingface_hub/hf_api.py
Comment (hlky) [reply src/huggingface_hub/hf_api.py]: Done, would you like a docstring as well?

Comment (Wauplin) [reply src/huggingface_hub/hf_api.py]: That would be great yes! Thanks!

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:9225]: Adding a small docstring would be great yes. Thanks in advance!

**Review by Wauplin — APPROVED** [review]
Thank you for the clear docstring! Looks ready to be merged! I'll just wait for the CI to pass before doing so :)

**Comment by hlky** [comment]
The only other thing I can think of is that there will be duplicate calls to the endpoint with `upload_folder` path, once during `_prepare_upload_folder_additions` then again during `create_commit`, if the extra load to the endpoint will be an issue there could be something like a boolean `self.yaml_validated` to skip the validation in `create_commit`.

**Comment by Wauplin** [comment]
Yeah, I thought the same as well - thanks for raising the question. Honestly, I don't think the extra overhead is really problematic. We could also have a small `@functools.cache` decorator on `_validate_yaml` (though it would be nice to cache only based on content+repo_type, not the token)

**Comment by hlky** [comment]
The fastai integration test failure appears to be because `repo_type` is `None`. It hasn't triggered before `repo_type` is [set to default value in `create_commit`](https://github.com/huggingface/huggingface_hub/blob/dfd73c05c7777804b14e9e3cec17e38b23c52ebe/src/huggingface_hub/hf_api.py#L3758). I'd suggest we duplicate that to `upload_folder`.

```python
import requests

content = '---\ntags:\n- fastai\n---\n\n# Amazing!\n\n🥳 Congratulations on hosting your fastai model on the Hugging Face Hub!\n\...n## Intended uses & limitations\nMore information needed\n\n## Training and evaluation data\nMore information needed\n'
url = "https://hub-ci.huggingface.co/api/validate-yaml"
r = requests.post(url, {"content": content, "repoType": None}) # {'error': '"value" does not match any of the allowed types'}

r = requests.post(url, {"content": content, "repoType": "model"}) # {'errors': [], 'warnings': []}
```

**Comment by Wauplin** [comment]
Thanks for looking into it. Yes, I would simply duplicate this line in `def _validate_yaml` as you mentioned. 

```py
repo_type = repo_type if repo_type is not None else REPO_TYPE_MODEL
```

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:4836]: I would move this to `_validate_yaml` directly to prevent errors in the future in case we use it somewhere else

Comment (hlky) [reply src/huggingface_hub/hf_api.py]: Thanks, I've moved it to `_validate_yaml`.

**Review by Wauplin — APPROVED** [review]

**Comment by Wauplin** [comment]
Perfect! Thanks for the iterations on this @hlky! Failing tests are unrelated to this PR so we are good to merge now! :tada:

**Pull request merged**
